pkg/i2gw/providers/openapi3/converter.go

```diff
@@ -0,0 +1,465 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi3
+
+import (
+	"fmt"
+	"regexp"
+	"slices"
+	"sort"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw"
+	"github.com/kubernetes-sigs/ingress2gateway/pkg/i2gw/providers/common"
+)
+
+const (
+	HostWildcard	 = "*"
+	HostSeparator  = ","
+	ParamSeparator = ","
+
+	HTTPRouteRulesMax      = 16
+	HTTPRouteMatchesMax    = 8
+	HTTPRouteMatchesMaxMax = HTTPRouteRulesMax * HTTPRouteMatchesMax
+)
+
+var uriRegexp = regexp.MustCompile(`^((https?)://([^/]+))?(/.*)?$`) // [_][1] = scheme, [_][2] = host, [_][3] = path
+
+type Converter interface {
+	Convert(Storage) (i2gw.GatewayResources, field.ErrorList)
+}
+
+// NewConverter returns a converter of OpenAPI Specifications 3.x from a storage into Gateway API resources.
+func NewConverter() Converter {
+	return &converter{}
+}
+
+type converter struct {}
+
+var _ Converter = &converter{}
+
+func (c *converter) Convert(storage Storage) (i2gw.GatewayResources, field.ErrorList) {
+	gatewayResources := i2gw.GatewayResources{
+		Gateways:        make(map[types.NamespacedName]gatewayv1.Gateway),
+		HTTPRoutes:      make(map[types.NamespacedName]gatewayv1.HTTPRoute),
+		TLSRoutes:       make(map[types.NamespacedName]gatewayv1alpha2.TLSRoute),
+		TCPRoutes:       make(map[types.NamespacedName]gatewayv1alpha2.TCPRoute),
+		ReferenceGrants: make(map[types.NamespacedName]gatewayv1beta1.ReferenceGrant),
+	}
```

Comment:
Do we really need to perform this initialization here?

Reply:
For `TLSRoutes`, `TCPRoutes` and `ReferenceGrants` probably not.

Nil maps are safe to read and it kinds of make sense to me that the caller will not try to write back on the `GatewayResources` struct without checking first. Well, if it does, it's the caller's problem anyway.

Besides, returning `nil` for `TLSRoutes`, `TCPRoutes` and `ReferenceGrants` seems semantically the right thing to do, I reckon. I'll fix it.

Thanks!